x/committee/client/common/query.go

```diff
@@ -0,0 +1,164 @@
+			// there should only be a single proposal under the given conditions
+			if msg.Type() == types.TypeMsgSubmitProposal {
+				subMsg := msg.(types.MsgSubmitProposal)
+				return NewProposer(proposalID, subMsg.Proposer.String()), nil
+			}
+		}
+	}
+
+	return Proposer{}, fmt.Errorf("failed to find the proposer for proposalID %d", proposalID)
+}
+
+// QueryProposalByID returns a proposal from state if present or fallbacks to searching old blocks
+func QueryProposalByID(cliCtx context.CLIContext, cdc *codec.Codec, queryRoute string, proposalID uint64) (*types.Proposal, int64, error) {
+	bz, err := cdc.MarshalJSON(types.NewQueryProposalParams(proposalID))
+	if err != nil {
+		return nil, 0, err
+	}
+
+	res, height, err := cliCtx.QueryWithData(fmt.Sprintf("custom/%s/%s", queryRoute, types.QueryProposal), bz)
+
+	if err == nil {
+		var proposal *types.Proposal
+		cdc.MustUnmarshalJSON(res, &proposal)
+
+		return proposal, height, nil
+	}
+
+	// NOTE: !errors.Is(err, types.ErrUnknownProposal) does not work here
+	if err != nil && !strings.Contains(err.Error(), "proposal not found") {
+		return nil, 0, err
+	}
+
+	res, height, err = cliCtx.QueryWithData(fmt.Sprintf("custom/%s/%s", queryRoute, types.QueryNextProposalID), nil)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	var nextProposalID uint64
+	cdc.MustUnmarshalJSON(res, &nextProposalID)
+
+	if proposalID >= nextProposalID {
+		return nil, 0, sdkerrors.Wrapf(types.ErrUnknownProposal, "%d", proposalID)
+	}
+
+	events := []string{
+		fmt.Sprintf("%s.%s='%s'", sdk.EventTypeMessage, sdk.AttributeKeyAction, types.TypeMsgSubmitProposal),
+		fmt.Sprintf("%s.%s='%s'", types.EventTypeProposalSubmit, types.AttributeKeyProposalID, []byte(fmt.Sprintf("%d", proposalID))),
+	}
+
+	searchResult, err := utils.QueryTxsByEvents(cliCtx, events, defaultPage, defaultLimit)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	for _, info := range searchResult.Txs {
+		for _, msg := range info.Tx.GetMsgs() {
+			if msg.Type() == types.TypeMsgSubmitProposal {
+				subMsg := msg.(types.MsgSubmitProposal)
+
+				deadline, err := calculateDeadline(cliCtx, cdc, queryRoute, subMsg.CommitteeID, info.Height)
+				if err != nil {
+					return nil, 0, err
+				}
+
+				return &types.Proposal{
+					ID:          proposalID,
+					CommitteeID: subMsg.CommitteeID,
+					PubProposal: subMsg.PubProposal,
+					Deadline:    deadline,
+				}, height, nil
+			}
+		}
+	}
+
+	return nil, 0, sdkerrors.Wrapf(types.ErrUnknownProposal, "%d", proposalID)
+}
+
+// calculateDeadline returns the proposal deadline for a committee and block height
+func calculateDeadline(cliCtx context.CLIContext, cdc *codec.Codec, queryRoute string, committeeID uint64, height int64) (time.Time, error) {
+	var deadline time.Time
+
+	bz, err := cdc.MarshalJSON(types.NewQueryCommitteeParams(committeeID))
+	if err != nil {
+		return deadline, err
+	}
+
+	res, height, err := cliCtx.QueryWithData(fmt.Sprintf("custom/%s/%s", queryRoute, types.QueryCommittee), bz)
```

Comment:
I'm getting the deadline calculated relative to the _current_ block time. I believe the variable `height` that was passed in to the function is getting overwritten here. I think you want to set the `cliCtx` to `cliCtx.WithHeight(height) and then do the query.